flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,591 @@
+
+                output.setRowKind(RowKind.UPDATE_AFTER);
+                output.replace(curRow, currFunction.getValue());
+                out.collect(output);
+            } else if (equaliser.equals(curRow, rowData)) {
+                // Retract record
+                output.setRowKind(RowKind.UPDATE_BEFORE);
+                output.replace(rowData, currFunction.getValue());
+                out.collect(output);
+                iterator.remove();
+                currFunction.retract(curRow);
+                isRetracted = true;
+            }
+        }
+
+        // Update state without the retracted row
+        inputState.update(rowList);
+    }
+
+    private void removeFromSortedListWithValueMap(RowData input, Collector<RowData> out)
+            throws Exception {
+        boolean isRetracted = false;
+        input.setRowKind(RowKind.INSERT);
+        List<RowData> sortedKeyList = sortedKeyState.value();
+        ListIterator<RowData> iterator = sortedKeyList.listIterator();
+
+        while (iterator.hasNext()) {
+            RowData curKey = iterator.next();
+            RowData curValue = valueMapState.get(curKey.getLong(keyIdx));
+            currFunction.accumulate(curValue);
+            prevFunction.accumulate(curValue);
+            if (isRetracted) {
+                // Emit updated agg value for all records after retraction
+                output.setRowKind(RowKind.UPDATE_BEFORE);
+                output.replace(curValue, prevFunction.getValue());
+                out.collect(output);
+
+                output.setRowKind(RowKind.UPDATE_AFTER);
+                output.replace(curValue, currFunction.getValue());
+                out.collect(output);
+            } else if (equaliser.equals(curValue, input)) {
+                // Retract record
+                output.setRowKind(RowKind.UPDATE_BEFORE);
+                output.replace(input, currFunction.getValue());
+                out.collect(output);
+                iterator.remove();
+                valueMapState.remove(curKey.getLong(keyIdx));
+                currFunction.retract(curValue);
+                isRetracted = true;
+            }
+        }
+
+        // Update sorted key state without the retracted row
+        sortedKeyState.update(sortedKeyList);
+    }
+
+    private void removeFromSortedListWithValueMapAndAccumulatorMap(
+            RowData input, Collector<RowData> out) throws Exception {
+        boolean isRetracted = false;
+        input.setRowKind(RowKind.INSERT);
+        List<RowData> sortedKeyList = sortedKeyState.value();
+        ListIterator<RowData> iterator = sortedKeyList.listIterator();
+
+        while (iterator.hasNext()) {
+            RowData curKey = iterator.next();
+            RowData curValue = valueMapState.get(curKey.getLong(keyIdx));
+            if (isRetracted) {
+                RowData prevAcc = accMapState.get(curKey.getLong(keyIdx));
+                if (prevAcc.equals(currFunction.getValue())) {
```

Comment:
We should compare the previous accumulator with the new accumulator (not the new value).
For the new value, we also need to `accumulate()` the current value first.

Reply:
Fixed